File: library/db/src/main/java/org/keynote/godtools/android/db/BaseMapper.java
package org.keynote.godtools.android.db;

import android.content.ContentValues;
import android.database.Cursor;
import android.support.annotation.NonNull;

import org.ccci.gto.android.common.db.AbstractMapper;
import org.cru.godtools.model.Base;

import static org.keynote.godtools.android.db.Contract.BaseTable.COLUMN_ID;

abstract class BaseMapper<T extends Base> extends AbstractMapper<T> {
    @Override
    protected void mapField(@NonNull final ContentValues values, @NonNull final String field, @NonNull final T obj) {
        switch (field) {
            case COLUMN_ID:
                values.put(field, obj.getId());
                break;
            default:
                super.mapField(values, field, obj);
                break;
        }
    }

    @NonNull
    @Override
    public T toObject(@NonNull final Cursor c) {
        final T obj = super.toObject(c);

        obj.setId(getLong(c, COLUMN_ID, Base.INVALID_ID));

        return obj;
    }
}

File: library/db/src/main/kotlin/org/keynote/godtools/android/db/BaseMapper.kt
package org.keynote.godtools.android.db

import android.content.ContentValues
import android.database.Cursor
import org.ccci.gto.android.common.db.AbstractMapper
import org.ccci.gto.android.common.util.database.getLong
import org.cru.godtools.model.Base
import org.keynote.godtools.android.db.Contract.BaseTable.Companion.COLUMN_ID

internal abstract class BaseMapper<T : Base> : AbstractMapper<T>() {
    override fun mapField(values: ContentValues, field: String, obj: T) {
        when (field) {
            COLUMN_ID -> values.put(field, obj.id)
            else -> super.mapField(values, field, obj)
        }
    }

    override fun toObject(c: Cursor): T = super.toObject(c).apply {
        id = c.getLong(COLUMN_ID, Base.INVALID_ID)
    }
}
